File: sdk/web/scripts/test-sqlite3mc-app-startup.py
"""Real application facade startup tests on this runner's synthetic OPFS only."""
import secrets


def run(ctx):
    js, snapshot, key, marker = (ctx[n] for n in ["js", "snapshot", "key", "marker"])
    checks = ctx["checks"]

    def load_app(mode=None, supplied=key):
        ctx["load"]()
        js("await new Promise((resolve,reject)=>{const script=document.createElement('script');script.type='module';script.src='/test-app-entry.js';script.onload=resolve;script.onerror=()=>reject(Error('app module failed to load'));document.head.append(script);});window.providerCalls=0;window.supplied=Uint8Array.from(arguments[1]);if(arguments[0])appFacade.configureStorageStartup({mode:arguments[0],keyProvider:async(id,purpose)=>{window.providerCalls++;if(id!=='spp.encrypted.db'||purpose!=='open')throw Error('provider context');return window.supplied;}});return true;", [mode, list(supplied)])

    def open_app():
        return js("const [a,b]=await Promise.all([appFacade.ensureStorage(),appFacade.ensureStorage()]);if(a!==b)throw Error('duplicate AppStorage');window.appStorage=a;return window.workerCount;")

    def refused():
        return js("try{await appFacade.ensureStorage();return false;}catch(e){window.startupError=String(e);return true;}")

    load_app()
    assert open_app() == 1
    assert js("return await appStorage.getSetting('integration-legacy');") == "legacy-preserved"
    js("await appStorage.setBootnodeConfig('https://synthetic.invalid');return true;")
    assert js("return await appStorage.getStoredBootnodeUrl();") == "https://synthetic.invalid"
    assert js("try{appFacade.configureStorageStartup({mode:'encrypted',keyProvider:()=>supplied});return false;}catch{return true;}")
    checks.append("app default startup shares one worker and preserves settings API")
    ctx["load"]()
    before = snapshot()

    if ctx["args"].plaintext_only:
        for mode in ["encrypted", "migrated"]:
            load_app(mode)
            assert refused()
            assert js("return startupError.includes('sqlite3mc feature')&&providerCalls===0&&workerCount===0;")
            assert snapshot() == before
            checks.append(f"stock app {mode} startup refuses before key provider without fallback")
        return

    ctx["encrypted"](True)
    js("await storage.call({SetSetting:{key:'app-protected',value_json:JSON.stringify(arguments[0])}});return true;", [marker])
    ctx["close"]()
    baseline = snapshot()
    for label, supplied in [("wrong key", secrets.token_bytes(32)), ("missing key", b"")]:
        load_app("encrypted", supplied)
        assert refused()
        assert snapshot() == baseline
        assert js("try{appFacade.configureStorageStartup({mode:'plaintext'});return false;}catch{return true;}")
        # Retry an unlocked provider without reconfiguring or creating a database.
        js("window.supplied=Uint8Array.from(arguments[0]);return true;", [list(key)])
        open_app()
        assert js("return await appStorage.getSetting('app-protected');") == marker
        checks.append(f"app {label} fails without writes or fallback; same-provider retry succeeds")
        ctx["load"]()
        baseline = snapshot()

    load_app("encrypted")
    assert open_app() == 1
    assert js("return window.providerCalls;") == 1
    js("await appStorage.setSetting('app-restart',arguments[0]);return true;", [marker])
    checks.append("app encrypted startup shares provider and worker")
    ctx["request"]("DELETE", "/session/" + ctx["sid"])
    ctx["sid"] = None
    ctx["session"]()
    load_app("encrypted")
    assert open_app() == 1
    assert js("return await appStorage.getSetting('app-restart');") == marker
    checks.append("app encrypted settings survive full browser restart")
    ctx["load"]()
    encrypted_files = [f for f in snapshot() if f["path"].startswith(".opfs-sahpool-encrypted/")]
    assert encrypted_files and not any(f.get("protected", False) for f in encrypted_files)
    assert [f for f in snapshot() if not f["path"].startswith(".opfs-sahpool-encrypted/")] == before
    checks.append("app encrypted writes preserve plaintext source and expose no protected marker")

    # Reuse only the isolated test profile; replace the encrypted fixture with
    # a new synthetic migration. The default source is still the original fixture.
    js("const root=await navigator.storage.getDirectory();await root.removeEntry('.opfs-sahpool-encrypted',{recursive:true});window.storage=await sdk.Storage.open();await storage.call({SetSetting:{key:'app-protected',value_json:JSON.stringify(arguments[0])}});return true;", [marker])
    ctx["close"]()

    def migration(action=None, create=False):
        ctx["load"]()
        return js("const m=await sdk.Storage.openMigration({createNew:arguments[1],keyProvider:()=>Uint8Array.from(arguments[0])});try{return arguments[2]?await m[arguments[2]]():await m.status();}finally{await m.close();}", [list(key), create, action])

    assert migration(create=True) == "copying"
    for status, action in [("copying", None), ("prepared", "prepare"), ("aborted", "abort")]:
        if action:
            assert migration(action) == status
        baseline = snapshot()
        load_app("migrated")
        assert refused()
        assert js("return startupError.includes(arguments[0]);", [status])
        assert snapshot() == baseline
        checks.append(f"app {status} migration refuses startup without fallback or data changes")

    assert migration("restart") == "copying"
    assert migration("prepare") == "prepared"
    assert migration("activate") == "active"
    baseline = snapshot()
    load_app("migrated", secrets.token_bytes(32))
    assert refused()
    assert snapshot() == baseline
    checks.append("app migrated startup authenticates control; wrong key leaves OPFS unchanged")
    load_app("migrated")
    assert open_app() == 2  # control worker released, then data worker
    assert js("return await appStorage.getSetting('app-protected');") == marker
    js("await appStorage.setSetting('app-after-activation',arguments[0]);return true;", [marker])
    load_app("migrated")
    assert open_app() == 2
    assert js("return await appStorage.getSetting('app-after-activation');") == marker
    checks.append("app active migration selects encrypted data and preserves post-activation writes")
    assert migration("finish") == "complete"
    load_app("migrated")
    assert open_app() == 2
    assert js("return await appStorage.getSetting('app-after-activation');") == marker
    ctx["load"]()
    assert not any(f.get("protected", False) for f in snapshot())
    checks.append("app completed migration reopens encrypted data after plaintext cleanup")

File: sdk/web/scripts/test-sqlite3mc-app-startup.js
import assert from "node:assert/strict";
import { randomBytes } from "node:crypto";

export async function run(ctx) {
  async function loadApp(mode = null, supplied = ctx.key) {
    await ctx.load();
    await ctx.js("await new Promise((resolve,reject)=>{const script=document.createElement('script');script.type='module';script.src='/test-app-entry.js';script.onload=resolve;script.onerror=()=>reject(Error('app module failed to load'));document.head.append(script);});window.providerCalls=0;window.supplied=Uint8Array.from(arguments[1]);if(arguments[0])appFacade.configureStorageStartup({mode:arguments[0],keyProvider:async(id,purpose)=>{window.providerCalls++;if(id!=='spp.encrypted.db'||purpose!=='open')throw Error('provider context');return window.supplied;}});return true;", [mode, Array.from(supplied)]);
  }
  const openApp = () => ctx.js("const [a,b]=await Promise.all([appFacade.ensureStorage(),appFacade.ensureStorage()]);if(a!==b)throw Error('duplicate AppStorage');window.appStorage=a;return window.workerCount;");
  const refused = () => ctx.js("try{await appFacade.ensureStorage();return false;}catch(e){window.startupError=String(e);return true;}");
  const snapshot = ctx.snapshot;

  await loadApp();
  assert.equal(await openApp(), 1);
  assert.equal(await ctx.js("return await appStorage.getSetting('integration-legacy');"), "legacy-preserved");
  await ctx.js("await appStorage.setBootnodeConfig('https://synthetic.invalid');return true;");
  assert.equal(await ctx.js("return await appStorage.getStoredBootnodeUrl();"), "https://synthetic.invalid");
  assert(await ctx.js("try{appFacade.configureStorageStartup({mode:'encrypted',keyProvider:()=>supplied});return false;}catch{return true;}"));
  ctx.checks.push("app default startup shares one worker and preserves settings API");
  await ctx.load();
  const before = await snapshot();
  if (ctx.args.plaintext_only) {
    for (const mode of ["encrypted", "migrated"]) {
      await loadApp(mode); assert(await refused());
      assert(await ctx.js("return startupError.includes('sqlite3mc feature')&&providerCalls===0&&workerCount===0;"));
      assert.deepEqual(await snapshot(), before);
      ctx.checks.push(`stock app ${mode} startup refuses before key provider without fallback`);
    }
    return;
  }
  await ctx.encrypted(true);
  await ctx.js("await storage.call({SetSetting:{key:'app-protected',value_json:JSON.stringify(arguments[0])}});return true;", [ctx.marker]);
  await ctx.close();
  let baseline = await snapshot();
  for (const [label, supplied] of [["wrong key", randomBytes(32)], ["missing key", Buffer.alloc(0)]]) {
    await loadApp("encrypted", supplied); assert(await refused());
    assert.deepEqual(await snapshot(), baseline);
    assert(await ctx.js("try{appFacade.configureStorageStartup({mode:'plaintext'});return false;}catch{return true;}"));
    await ctx.js("window.supplied=Uint8Array.from(arguments[0]);return true;", [Array.from(ctx.key)]);
    await openApp();
    assert.equal(await ctx.js("return await appStorage.getSetting('app-protected');"), ctx.marker);
    ctx.checks.push(`app ${label} fails without writes or fallback; same-provider retry succeeds`);
    await ctx.load(); baseline = await snapshot();
  }
  await loadApp("encrypted");
  assert.equal(await openApp(), 1);
  assert.equal(await ctx.js("return window.providerCalls;"), 1);
  await ctx.js("await appStorage.setSetting('app-restart',arguments[0]);return true;", [ctx.marker]);
  ctx.checks.push("app encrypted startup shares provider and worker");
  await ctx.request("DELETE", `/session/${ctx.sid}`); ctx.sid = null;
  await ctx.session(); await loadApp("encrypted");
  assert.equal(await openApp(), 1);
  assert.equal(await ctx.js("return await appStorage.getSetting('app-restart');"), ctx.marker);
  ctx.checks.push("app encrypted settings survive full browser restart");
  await ctx.load();
  const encryptedFiles = (await snapshot()).filter(file => file.path.startsWith(".opfs-sahpool-encrypted/"));
  assert(encryptedFiles.length && !encryptedFiles.some(file => file.protected));
  assert.deepEqual((await snapshot()).filter(file => !file.path.startsWith(".opfs-sahpool-encrypted/")), before);
  ctx.checks.push("app encrypted writes preserve plaintext source and expose no protected marker");
  await ctx.js("const root=await navigator.storage.getDirectory();await root.removeEntry('.opfs-sahpool-encrypted',{recursive:true});window.storage=await sdk.Storage.open();await storage.call({SetSetting:{key:'app-protected',value_json:JSON.stringify(arguments[0])}});return true;", [ctx.marker]);
  await ctx.close();
  async function migration(action = null, create = false) {
    await ctx.load();
    return ctx.js("const m=await sdk.Storage.openMigration({createNew:arguments[1],keyProvider:()=>Uint8Array.from(arguments[0])});try{return arguments[2]?await m[arguments[2]]():await m.status();}finally{await m.close();}", [Array.from(ctx.key), create, action]);
  }
  assert.equal(await migration(null, true), "copying");
  for (const [status, action] of [["copying", null], ["prepared", "prepare"], ["aborted", "abort"]]) {
    if (action) assert.equal(await migration(action), status);
    baseline = await snapshot(); await loadApp("migrated"); assert(await refused());
    assert(await ctx.js("return startupError.includes(arguments[0]);", [status]));
    assert.deepEqual(await snapshot(), baseline);
    ctx.checks.push(`app ${status} migration refuses startup without fallback or data changes`);
  }
  assert.equal(await migration("restart"), "copying");
  assert.equal(await migration("prepare"), "prepared");
  assert.equal(await migration("activate"), "active");
  baseline = await snapshot(); await loadApp("migrated", randomBytes(32)); assert(await refused());
  assert.deepEqual(await snapshot(), baseline);
  ctx.checks.push("app migrated startup authenticates control; wrong key leaves OPFS unchanged");
  await loadApp("migrated"); assert.equal(await openApp(), 2);
  assert.equal(await ctx.js("return await appStorage.getSetting('app-protected');"), ctx.marker);
  await ctx.js("await appStorage.setSetting('app-after-activation',arguments[0]);return true;", [ctx.marker]);
  await loadApp("migrated"); assert.equal(await openApp(), 2);
  assert.equal(await ctx.js("return await appStorage.getSetting('app-after-activation');"), ctx.marker);
  ctx.checks.push("app active migration selects encrypted data and preserves post-activation writes");
  assert.equal(await migration("finish"), "complete");
  await loadApp("migrated"); assert.equal(await openApp(), 2);
  assert.equal(await ctx.js("return await appStorage.getSetting('app-after-activation');"), ctx.marker);
  await ctx.load(); assert(!(await snapshot()).some(file => file.protected));
  ctx.checks.push("app completed migration reopens encrypted data after plaintext cleanup");
}
